packages/xo-server/src/xo-mixins/pool.js

```diff
@@ -0,0 +1,93 @@
+import { difference } from 'lodash'
+
+export default class {
+  constructor(xo) {
+    this._xo = xo
+  }
+
+  async mergeInto({ sources, target, force }) {
+    const targetHost = this._xo.getObject(target.master)
+    const _sources = sources.map(sourceId =>
+      this._xo.getObject(sourceId, 'pool')
+    )
+    let source
+    let sourceHost
+    const patchesSources = {}
+    for (source of _sources) {
+      sourceHost = this._xo.getObject(source.master)
+      if (sourceHost.productBrand !== targetHost.productBrand) {
+        throw new Error(
+          `a ${sourceHost.productBrand} pool cannot be merged into a ${
+            targetHost.productBrand
+          } pool`
+        )
+      }
+      patchesSources[source.id] = sourceHost.patches.map(
+        patchId => this._xo.getObject(patchId).name
+      )
+    }
+
+    const requiredPatchesSources = {}
+    let requiredPatchesTarget = []
+    for (source of _sources) {
+      requiredPatchesTarget = requiredPatchesTarget.concat(
+        this._xo.getPatchesDifference(source.master, target.master)
+      )
+      requiredPatchesSources[source.id] = this._xo.getPatchesDifference(
+        target.master,
+        source.master
+      )
+    }
+
+    requiredPatchesTarget = [...new Set(requiredPatchesTarget)]
+    const newPatchesTarget = requiredPatchesTarget.concat(
+      targetHost.patches.map(patchId => this._xo.getObject(patchId).name)
+    )
+    let newPatchesSource
+    let requiredPatchesSource
+    for (source of _sources) {
+      requiredPatchesSource = requiredPatchesSources[source.id]
+      newPatchesSource = requiredPatchesSource.concat(patchesSources[source.id])
+      requiredPatchesTarget = requiredPatchesTarget.concat(
+        difference(newPatchesSource, newPatchesTarget)
+      )
+      requiredPatchesSources[source.id] = requiredPatchesSource.concat(
+        difference(newPatchesTarget, newPatchesSource)
+      )
+    }
+
+    if (requiredPatchesTarget.length > 0) {
+      const targetXapi = this._xo.getXapi(target)
+      await targetXapi.installPatches({
+        patches: await targetXapi.findPatches(requiredPatchesTarget),
+      })
+    }
+
+    let sourceXapi
+    await Promise.all(
+      _sources.map(async source => {
+        if (requiredPatchesSources[source.id].length > 0) {
+          sourceXapi = this._xo.getXapi(source)
+          return sourceXapi.installPatches({
+            patches: await sourceXapi.findPatches(
+              requiredPatchesSources[source.id]
+            ),
+          })
+        }
+      })
+    )
```

Comment:
Ok, first let's make sure we agree on the algorithm:
- create a map of { pool ID → patches to be installed on the pool }
- create a set of patches
- for each source, add (source - target) to the set
- add the content of the set to the map (target)
- add the target's patches to the set
- for each source, add (set - source) to the map
- install all the patches

Don't bother changing anything until we agree on those steps :)